### PR TITLE
[Bug-fix][XLA:CPU][oneDNN] Fix BINARY_ADD fusion to Dot

### PR DESCRIPTION
**Pull request by mdfaijul**
This PR fixes a bug reported for JAX (https://github.com/openxla/xla/issues/13054)

### CODE REVIEW

File: xla/tests/onednn_matmul_test.cc [hunk 1]
```diff
@@ -1442,6 +1442,52 @@ TEST_F(MatmulTest, WeightsPrepackAndScratch) {
...
+
+  region_1.29 {
+    Arg_0.30 = f32[] parameter(0)
+    Arg_1.31 = f32[] parameter(1)
+    ROOT add.32 = f32[] add(Arg_0.30, Arg_1.31)
+  }
+
+  ENTRY main {
+    constant.2 = f32[] constant(1e-06)
+    broadcast.3 = f32[1000000] broadcast(constant.2), dimensions={}
```
Comment (penpornk) [inline xla/tests/onednn_matmul_test.cc:1462]: I don't think the size needs to be this big to reproduce the failure. Would 10 work?

File: xla/tests/onednn_matmul_test.cc [hunk 2]
```diff
@@ -1442,6 +1442,52 @@ TEST_F(MatmulTest, WeightsPrepackAndScratch) {
...
+    broadcast.8 = f32[1000000,3] broadcast(constant.7), dimensions={}
+    Arg_0.1 = f32[3] parameter(0)
+    reshape.10 = f32[1,3] reshape(Arg_0.1)
+    broadcast.11 = f32[1,3] broadcast(reshape.10), dimensions={0,1}
+    reshape.12 = f32[3] reshape(broadcast.11)
+    broadcast.13 = f32[1000000,3] broadcast(reshape.12), dimensions={1}
+    subtract.14 = f32[1000000,3] subtract(broadcast.8, broadcast.13)
+    constant.4 = f32[] constant(0)
+    broadcast.5 = f32[3,3] broadcast(constant.4), dimensions={}
+    dot.15 = f32[1000000,3] dot(subtract.14, broadcast.5), lhs_contracting_dims={1}, rhs_contracting_dims={0}
```
Comment (penpornk) [inline xla/tests/onednn_matmul_test.cc:1473]: Can we reduce the ops to just necessary ops that reproduce the failure? I don't think all the dots are needed.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the fix! Could you please explain what is causing the issue, and how this fix addresses it? Is it because of rank mismatch + wrong auto broadcasting or something?

File: xla/tests/onednn_matmul_test.cc (hunk 2 above)
Comment (kanvi-nervana) [reply xla/tests/onednn_matmul_test.cc:1473]: The bug seems to be seen with this particular case.

**Comment by kanvi-nervana** [comment]
> Thank you for the fix! Could you please explain what is causing the issue, and how this fix addresses it? Is it because of rank mismatch + wrong auto broadcasting or something?

oneDNN expects Matmul followed by Bias-Add followed by Binary-Add. But, here Matmul is followed by Binary-Add and then by Bias-Add which oneDNN does not support. The fix here is extending the dimensions of the Bias-Add to a Binary-Add which is supported. As seen below

Screenshot 2024-06-03 at 4 02 50 PM

File: xla/tests/onednn_matmul_test.cc (hunk 1 above)
Comment (kanvi-nervana) [reply xla/tests/onednn_matmul_test.cc:1462]: The issue is not reproducible with a smaller size.

**Review by penpornk — APPROVED** [review]
Thank you very much for the clarifications!

**Pull request closed without merging**